packages/doenetml-worker-rust/lib-doenetml-core/src/core/state/state_var/views.rs

```diff
@@ -0,0 +1,325 @@
+        // Note: slower than it seems necessary due to two borrows.
+        // Another option is to use the ref_filter_map crate or see if there will eventually be a way
+        // to convert Ref<Option<T>> to Option<Ref<T>>.
+        // See: https://stackoverflow.com/a/62474266
+        if self.inner.borrow().try_get_last_value().is_none() {
+            None
+        } else {
+            Some(Ref::map(self.inner.borrow(), |v| {
+                v.try_get_last_value().unwrap()
+            }))
+        }
+    }
+
+    /// Set the value of the state variable to the supplied value,
+    /// set `came_from_default` to false, and mark it fresh
+    pub fn set_value(&self, new_val: T) {
+        self.inner.borrow_mut().set_value(new_val);
+    }
+
+    /// Set the value of the state variable to `new_val`,
+    /// mark it as Fresh, and set `came_from_default` to true.
+    pub fn set_value_and_set_came_from_default(&self, new_val: T) {
+        self.inner
+            .borrow_mut()
+            .set_value_and_set_came_from_default(new_val);
+    }
+
+    /// If the state variable is Stale, mark it as Fresh
+    /// so that the value it had before `mark_stale` was called
+    /// will be its fresh value again.
+    ///
+    /// Panics: if the state variable is Unresolved.
+    pub fn mark_fresh(&self) {
+        self.inner.borrow_mut().mark_fresh();
+    }
+
+    /// Return if the value of this state variable was set from its default value
+    pub fn came_from_default(&self) -> bool {
+        self.inner.borrow().came_from_default()
+    }
+
+    /// Set the freshness of the variable to Stale
+    pub fn mark_stale(&self) {
+        self.inner.borrow_mut().mark_stale()
+    }
+
+    /// Set the freshness of the variable to Resolved
+    pub fn set_as_resolved(&self) {
+        self.inner.borrow_mut().set_as_resolved()
+    }
+
+    /// Return the current freshness of the variable
+    pub fn get_freshness(&self) -> Freshness {
+        self.inner.borrow().freshness
+    }
+
+    /// Set the `requested_value` field to the supplied value
+    pub fn set_requested_value(&self, requested_val: T) {
+        self.inner.borrow_mut().set_requested_value(requested_val);
+    }
+
+    /// Get a reference to the current `requested_value` field
+    pub fn get_requested_value(&self) -> impl Deref<Target = T> + '_ {
+        Ref::map(self.inner.borrow(), |v| v.get_requested_value())
+    }
+
+    /// Set the value of the state variable to the current value of its `requested_value` field.
+    ///
+    /// Equivalent to calling `get_requested_value()` and then calling `set_value()` with that value,
+    /// except more efficient due to only one borrow of the inner value.
+    pub fn set_value_to_requested_value(&self) {
+        let mut inner = self.inner.borrow_mut();
+        let new_value = inner.get_requested_value().clone();
+        inner.set_value(new_value);
+    }
+}
+
+impl<T: Default + Clone> StateVarView<T> {
```

Comment:
If they're in the same file, can we group the struct def and the impl together?

Reply:
Done